From 0d1e62b1f5cf07ef38b728808664abcfe54e5ac3 Mon Sep 17 00:00:00 2001
From: Christophe Guibout <christophe.guibout@st.com>
Date: Wed, 12 Jun 2024 11:11:08 +0200
Subject: [PATCH 1/3] autoboot: propagate boot index from tf-a

tf-a reads boot index from metadata, and write it into FWU
backup register.
During an update, in case of failure of active partition after
several attempts (bootcount max count triggered), the system
needs to reboot on previous_active partition, so u-boot needs
to read FWU_INFO backup register (fwu metadata partition is
not updated by tf-a).
fwu_plat_get_bootidx() function reads the backup register to
load bootcmd or altbootcmd.
---
 arch/arm/mach-stm32mp/include/mach/stm32.h |  6 +++++-
 board/st/stm32mp2/stm32mp2.c               | 20 ++++++++++++++++++++
 common/autoboot.c                          |  4 +++-
 3 files changed, 28 insertions(+), 2 deletions(-)

diff --git a/arch/arm/mach-stm32mp/include/mach/stm32.h b/arch/arm/mach-stm32mp/include/mach/stm32.h
index 7c4836a894..1c5df49b65 100644
--- a/arch/arm/mach-stm32mp/include/mach/stm32.h
+++ b/arch/arm/mach-stm32mp/include/mach/stm32.h
@@ -138,7 +138,6 @@ enum forced_boot_mode {
 #define TAMP_BOOTCOUNT			TAMP_BACKUP_REGISTER(21)
 
 #define TAMP_FWU_BOOT_IDX_MASK		GENMASK(3, 0)
-
 #define TAMP_FWU_BOOT_IDX_OFFSET	0
 #endif
 
@@ -185,8 +184,13 @@ enum forced_boot_mode {
 /* TAMP registers x = 0 to 127 : hardcoded description, waiting NVMEM node in DT */
 #define TAMP_BACKUP_REGISTER(x)		(STM32_TAMP_BASE + 0x100 + 4 * (x))
 
+#define TAMP_FWU_BOOT_INFO_REG		TAMP_BACKUP_REGISTER(48)
+
 /* TAMP registers zone 3 RIF 1 (RW) at 96*/
 #define TAMP_BOOT_CONTEXT		TAMP_BACKUP_REGISTER(96)
+
+#define TAMP_FWU_BOOT_IDX_MASK		GENMASK(3, 0)
+#define TAMP_FWU_BOOT_IDX_OFFSET	0
 #endif /* defined(CONFIG_STM32MP21X) || defined(CONFIG_STM32MP23X) || defined(CONFIG_STM32MP25X) */
 
 /* offset used for BSEC driver: misc_read and misc_write */
diff --git a/board/st/stm32mp2/stm32mp2.c b/board/st/stm32mp2/stm32mp2.c
index a8b1cdfbaa..e2a9f020f3 100644
--- a/board/st/stm32mp2/stm32mp2.c
+++ b/board/st/stm32mp2/stm32mp2.c
@@ -891,3 +891,23 @@ int board_fix_fdt(void *blob)
 }
 #endif /* CONFIG_OF_BOARD_FIXUP */
 
+#if defined(CONFIG_FWU_MULTI_BANK_UPDATE)
+
+#include <fwu.h>
+
+/**
+ * fwu_plat_get_bootidx() - Get the value of the boot index
+ * @boot_idx: Boot index value
+ *
+ * Get the value of the bank(partition) from which the platform
+ * has booted. This value is passed to U-Boot from the earlier
+ * stage bootloader which loads and boots all the relevant
+ * firmware images
+ *
+ */
+void fwu_plat_get_bootidx(uint *boot_idx)
+{
+	*boot_idx = (readl(TAMP_FWU_BOOT_INFO_REG) >>
+		    TAMP_FWU_BOOT_IDX_OFFSET) & TAMP_FWU_BOOT_IDX_MASK;
+}
+#endif /* CONFIG_FWU_MULTI_BANK_UPDATE */
diff --git a/common/autoboot.c b/common/autoboot.c
index 5d331991c1..970706af20 100644
--- a/common/autoboot.c
+++ b/common/autoboot.c
@@ -447,6 +447,7 @@ const char *bootdelay_process(void)
 {
 	char *s;
 	int bootdelay;
+	u32 boot_idx;
 
 	bootcount_inc();
 
@@ -472,7 +473,8 @@ const char *bootdelay_process(void)
 		s = env_get("failbootcmd");
 	} else
 #endif /* CONFIG_POST */
-	if (bootcount_error())
+	fwu_plat_get_bootidx(&boot_idx);
+	if (boot_idx)
 		s = env_get("altbootcmd");
 	else
 		s = env_get("bootcmd");
-- 
2.34.1

